File: suites/01_algorithms/comparisons/quicksort.go
package main

import "fmt"

func partitionCost(low, high, pivot int64) int64 {
	var score int64
	for i := low; i < high; i++ {
		probe := ((i * 17) + 13) % 997
		if probe < pivot {
			score += probe
		} else {
			score -= probe
		}
	}
	return score
}

func quicksortCost(low, high int64) int64 {
	if high-low < 2 {
		return high - low
	}
	pivot := (low + high) / 2
	return partitionCost(low, high, pivot) +
		quicksortCost(low, pivot) +
		quicksortCost(pivot+1, high)
}

func main() {
	fmt.Println(quicksortCost(0, 5000))
}

File: suites/01_algorithms/comparisons/quicksort.rs
fn partition_cost(low: i64, high: i64, pivot: i64) -> i64 {
    let mut score: i64 = 0;
    for i in low..high {
        let probe = ((i * 17) + 13) % 997;
        if probe < pivot {
            score += probe;
        } else {
            score -= probe;
        }
    }
    score
}

fn quicksort_cost(low: i64, high: i64) -> i64 {
    if high - low < 2 {
        return high - low;
    }
    let pivot = (low + high) / 2;
    partition_cost(low, high, pivot)
        + quicksort_cost(low, pivot)
        + quicksort_cost(pivot + 1, high)
}

fn main() {
    println!("{}", quicksort_cost(0, 5000));
}
